### PR TITLE
[data] New executor [10/n]--- Plumbing for locality_with_output and setting execution options

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?
1. Support locality_with_output.
2. Support setting execution options via DatasetContext for now (and a TODO to add it to the public API).
3. Add unit test for configuration plumbing.

### CODE REVIEW

File: python/ray/data/_internal/plan.py [hunk 1]
```diff
@@ -449,12 +449,11 @@ def execute_to_iterator(
             )
 
         from ray.data._internal.execution.streaming_executor import StreamingExecutor
-        from ray.data._internal.execution.interfaces import ExecutionOptions
         from ray.data._internal.execution.legacy_compat import (
             execute_to_legacy_block_iterator,
         )
 
-        executor = StreamingExecutor(ExecutionOptions())
+        executor = StreamingExecutor(copy.deepcopy(ctx.execution_options))
```
Comment (c21) [inline python/ray/data/_internal/plan.py:456]: given `execution_options` could be muted in `StreamingExecutor`/`BulkExecutor`, shall we just provide a method in `DatasetContext.get_execution_options()` to just return a deep copy of `ctx.execution_options`?

**Review by c21 — APPROVED** [review]
LGTM w/ minor comment.

File: python/ray/data/context.py [hunk 2]
```diff
@@ -171,6 +175,8 @@ def __init__(
         self.enable_auto_log_stats = enable_auto_log_stats
         self.trace_allocations = trace_allocations
         self.optimizer_enabled = optimizer_enabled
+        # TODO: expose execution options in Dataset public APIs.
```
Comment (jianoaix) [inline python/ray/data/context.py:178]: It'd be better to create a tracking issue

**Review by jianoaix — APPROVED** [review]

**Review by jianoaix — APPROVED** [review]

File: python/ray/data/_internal/plan.py (hunk 1 above)
Comment (ericl) [reply python/ray/data/_internal/plan.py:456]: Missed this comment--- I'll file a followup.

File: python/ray/data/context.py (hunk 2 above)
Comment (ericl) [reply python/ray/data/context.py:178]: Added to https://github.com/ray-project/ray/issues/31797

**Pull request merged**